### PR TITLE
feat(metrics): Integrate new metrics layer into the metrics API

### LINKED ISSUE #57539: Start using the new metrics layer

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR introduces the basic integration of the new metrics layer in the metrics API. This implementation serves as of now an experimental purpose, to explore the capability of the new layer.

In the future we plan to further improve and refine the implementation, once more features will be pushed down and added to the metrics layer.

Closes https://github.com/getsentry/sentry/issues/57539

### CODE REVIEW

File: src/sentry/snuba/metrics_layer/api.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+    intervals = [datetime.fromtimestamp(current_time, timezone.utc)]
+    while current_time + interval < end_seconds:
+        next_time = current_time + interval
+        intervals.append(datetime.fromtimestamp(next_time, timezone.utc))
+        current_time = next_time
+
+    return intervals
+
+
+def _zerofill_series(
```
Comment (evanh) [inline src/sentry/snuba/metrics_layer/api.py:188]: Is this all that the zerofill needed to do? I imagined it would build a row with all the same keys, but a zero value. E.g., A query that wants a row for time 1, 2, 3, but there is no data for time 2:

```
[
   {"time": 1, "aggregate_value": 65},
   {"time": 3, "aggregate_value": 45},
]
```

Note a missing row for `"time": 2`. What I thought the zerofill would do is add a proper row value, so the final result would be:

```
[
   {"time": 1, "aggregate_value": 65},
   {"time": 2, "aggregate_value": 0},    <- new row
   {"time": 3, "aggregate_value": 45},
]
```

File: src/sentry/snuba/metrics_layer/api.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
+import re
```
Comment (evanh) [inline src/sentry/snuba/metrics_layer/api.py:1]: We talked about this in Slack, but I don't think this belongs in the `snuba` folder. Part of this is a naming issue we should resolve.

Metrics API: The external metrics API
Existing Metrics Layer: The metrics layer that uses DerivedMetrics etc.
New Metrics Layer/API?: The new layer/API that Enoch and I are working on. I was calling it the metrics API to distinguish it from the metrics layer above but now that is confusing.

I think it's important to keep the hierarchy clear: The API calls the new layer. The new layer is also not specific to the API (it will be used by all features that use metrics). All the code for this API should live outside of the snuba folder.

File: src/sentry/snuba/metrics_layer/api.py (hunk 1 above)
Comment (iambriccardo) [reply src/sentry/snuba/metrics_layer/api.py:188]: Yep, initially it was filling zero elements but now I put `None` since this was the old API behavior from my testing. I will have to change the name of the function.

This algorithm does zerofilling in the way you said, but in the inverse, it builds all the intervals that are requested and defaults them to `None` (but can easily be `0`) and then for each returned value it substitutes the `None` with the actual value.

File: src/sentry/snuba/metrics_layer/api.py (hunk 2 above)
Comment (iambriccardo) [reply src/sentry/snuba/metrics_layer/api.py:1]: Yep! I will definitely find a better place for it

File: src/sentry/ddm/api.py [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+    10,  # 10 seconds
+    60,  # 1 minute
+    60 * 60,  # 1 hour
+    60 * 60 * 24,  # 24 hours
+]
+
+# These regexes are temporary since the DSL is supposed to be parsed internally by the snuba SDK, thus this
+# is only bridging code to validate and evolve the metrics layer.
+FIELD_REGEX = re.compile(r"^(\w+)\(([^\s]+)\)$")
+QUERY_REGEX = re.compile(r"(\w+):([^\s]+)(?:\s*)")
```
Comment (jan-auer) [inline src/sentry/ddm/api.py:35]:
```suggestion
QUERY_REGEX = re.compile(r"(\w+):([^\s]+)")
```

File: src/sentry/ddm/api.py [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+GRANULARITIES = [
+    10,  # 10 seconds
+    60,  # 1 minute
+    60 * 60,  # 1 hour
+    60 * 60 * 24,  # 24 hours
+]
+
+# These regexes are temporary since the DSL is supposed to be parsed internally by the snuba SDK, thus this
+# is only bridging code to validate and evolve the metrics layer.
+FIELD_REGEX = re.compile(r"^(\w+)\(([^\s]+)\)$")
```
Comment (jan-auer) [inline src/sentry/ddm/api.py:34]:
```suggestion
FIELD_REGEX = re.compile(r"^(\w+)\(([^\s)]+)\)$")
```

File: src/sentry/ddm/api.py
Comment (jan-auer) [inline src/sentry/ddm/api.py:1]: Let's move this to `sentry.sentry_metrics.api`

File: src/sentry/ddm/api.py [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+    dataset = Dataset.Metrics if use_case_id == UseCaseID.SESSIONS else Dataset.PerformanceMetrics
+    request = Request(
+        dataset=dataset.value,
+        query=base_query,
+        app_id="default",
+        tenant_ids={"referrer": "metrics.data", "organization_id": organization.id},
+    )
+
+    base_query.rollup = Rollup(interval=interval, granularity=_get_granularity(interval))
+    series_result = run_query(request=request)
```
Comment (jan-auer) [inline src/sentry/ddm/api.py:318]: Let's run a single query with `totals=True` -- as long as there are bugs, we can temporarly disable that but rather not workaround it. Please ask to prioritize the fix, instead.

File: src/sentry/ddm/api.py [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+            org_ids=[organization.id],
+            project_ids=[project.id for project in projects],
+            use_case_id=use_case_id.value,
+        ),
+    )
+
+    # Parse all the strings and convert them to an intermediate format.
+    parsed_fields = _parse_fields(fields)
+    parsed_filters = _parse_query(query)
+    parsed_group_bys = _parse_group_by(group_bys)
```
Comment (jan-auer) [inline src/sentry/ddm/api.py:360]: Please parse directly into the metrics query types, which should allow us to shorten this similar to:

```py
    base_query = MetricsQuery(
        filters=_parse_filters(query) if query else None,
        groupby=[Column(g) for g in group_bys] if group_bys else None,
        start=start,
        end=end,
        scope=...,
    )

    for timeseries in _parse_fields(fields):
        result = run_query(base_query.set_query(timeseries))
        ...
```

File: src/sentry/api/endpoints/organization_metrics.py [hunk 7]
```diff
@@ -154,7 +157,31 @@ class OrganizationMetricsDataEndpoint(OrganizationEndpoint):
...
+        # We then run the query and inject directly the field, query and groupBy, since they will be parsed
+        # internally.
+        results = run_metrics_query(
+            fields=request.GET.getlist("field", []),
+            query=request.GET.get("query"),
+            group_bys=request.GET.getlist("groupBy"),
+            interval=interval,
+            start=start,
+            end=end,
+            use_case_id=get_use_case_id(request),
```
Comment (jan-auer) [inline src/sentry/api/endpoints/organization_metrics.py:177]: Do not pass the `use_case_id` anymore, instead infer it _internally_ from the parsed `Timeseries` since the MRIs by design specify the namespace (aka use case). This should eventually be handled by the query layer.

File: src/sentry/snuba/metrics/query_builder.py [hunk 8]
```diff
@@ -501,7 +501,7 @@ def __init__(
             parse_field(
                 key,
                 allow_mri=allow_mri,
-                allow_private=bool(query_params.get("allowPrivate", False)),
+                allow_private=query_params.get("allowPrivate", "false") == "true",
```
Comment (jan-auer) [inline src/sentry/snuba/metrics/query_builder.py:504]: Seems unrelated to this change.

File: src/sentry/sentry_metrics/querying/api.py [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+    )
+
+
+def _execute_series_and_totals_query(
+    organization: Organization, use_case_id: UseCaseID, interval: int, base_query: MetricsQuery
+) -> Mapping[str, Any]:
+    query = base_query.set_rollup(Rollup(interval=interval, granularity=_get_granularity(interval)))
+    series_result = run_query(request=_build_request(organization, use_case_id, query))
+
+    # TODO(layer): The layer should be able to make the totals query and the series query in one.
```
Comment (evanh) [inline src/sentry/sentry_metrics/querying/api.py:256]: > The layer should be able to make the totals query and the series query in one.

Can you elaborate on what you mean here? Why would the layer make two queries?

Comment (evanh) [reply src/sentry/sentry_metrics/querying/api.py:256]: Is this function fetching the timeseries and the totals for the same timeseries? That's something Clickhouse supports automatically (see the `WITH TOTALS` [clause](https://clickhouse.com/docs/en/sql-reference/statements/select/group-by#with-totals-modifier)). Is that what you're trying to do here?

File: src/sentry/sentry_metrics/querying/api.py [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+def _build_request(
+    organization: Organization, use_case_id: UseCaseID, query: MetricsQuery
+) -> Request:
+    # TODO(layer): The layer should handle the dataset selection automatically.
+    dataset = Dataset.Metrics if use_case_id == UseCaseID.SESSIONS else Dataset.PerformanceMetrics
+    return Request(
+        dataset=dataset.value,
+        query=query,
+        app_id="default",
+        tenant_ids={"referrer": "metrics.data", "organization_id": organization.id},
```
[the remark this answers is not in the document]
Comment (iambriccardo) [reply src/sentry/sentry_metrics/querying/api.py:246]: This is a great suggestion but we might have to see how to infer the consumer, since we could do this via a query param. @jan-auer wdyt?

File: src/sentry/sentry_metrics/querying/api.py (hunk 9 above)
Comment (iambriccardo) [reply src/sentry/sentry_metrics/querying/api.py:256]: Exactly

Comment (evanh) [reply src/sentry/sentry_metrics/querying/api.py:256]: OK well then we can support adding the `WITH TOTALS` value to the query. That should return the totals in a single query.

File: src/sentry/sentry_metrics/querying/api.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (evanh) [reply src/sentry/sentry_metrics/querying/api.py:246]: If this API is going to sit on the ingestion path then I definitely recommend calling that out separate somehow. We definitely want to throttle that use case in a different way from everything else.

File: src/sentry/sentry_metrics/querying/api.py [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+        filters=_parse_filters(query) if query else None,
+        groupby=[Column(group_by) for group_by in group_bys] if group_bys else None,
+        start=start,
+        end=end,
+    )
+
+    # For each field generate the query.
+    query_results = []
+    for timeseries in _parse_fields(fields):
+        # TODO(layer): The layer should automatically infer the use case id from the passed metrics.
```
[the remark this answers is not in the document]
Comment (evanh) [reply src/sentry/sentry_metrics/querying/api.py:298]: We resolved the debate lol. The layer will infer the use case ID and granularity automatically.

File: src/sentry/sentry_metrics/querying/api.py [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+        return extract_use_case_id(mri)
+    except ValidationError:
+        raise InvalidParams(f"The query contains an invalid MRI: {mri}")
+
+
+def _execute_series_and_totals_query(
+    organization: Organization, interval: int, base_query: MetricsQuery
+) -> Mapping[str, Any]:
+    # We infer the use case id from the timeseries.
+    inferred_use_case_id = _infer_use_case_id_from_timeseries(base_query.query)
```
Comment (evanh) [inline src/sentry/sentry_metrics/querying/api.py:246]: Is this still necessary if the layer is also inferring the use case ID?

File: src/sentry/sentry_metrics/querying/api.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (evanh) [reply src/sentry/sentry_metrics/querying/api.py:226]: Any more thoughts on this?

File: src/sentry/sentry_metrics/querying/api.py [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+    projects: Sequence[Project],
+):
+    base_scope = MetricsScope(
+        org_ids=[organization.id],
+        project_ids=[project.id for project in projects],
+    )
+
+    # Build the basic query that contains the metadata.
+    base_query = MetricsQuery(
+        filters=_parse_filters(query) if query else None,
```
Comment (evanh) [inline src/sentry/sentry_metrics/querying/api.py:282]: Enoch has removed the filters and groupby at the `MetricsQuery` layer in a future version of the SDK. So we will only support them on the `Timeseries` directly. If its possible to incorporate that now, it will make our lives easier in the future.

**Review by evanh — APPROVED** [review]
Couple questions/comments. But generally looks good to me!

File: src/sentry/sentry_metrics/querying/api.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jan-auer) [reply src/sentry/sentry_metrics/querying/api.py:226]: This API isn't going to be on the ingestion path. For now, its only use is the organization metrics data endpoint; though I envision this to become a common base class for many metrics-based APIs. In that case, I think the API endpoint class would set/override the referrer.

We can prepare for this now already and pass the referrer in from OrganizationMetricsDataEndpoint.

**Review by jan-auer — APPROVED** [review]
Good to go, once these two items are addressed:
1. referrer moved into the endpoint
2. filters/group-by moved into timeseries 

Regarding use case inference, it's fine to merge as-is and follow up later. It can be removed with Snuba inferring the dataset; we should still not pass session queries through the new layer since we don't support the derived metrics for sessions yet.

Please then follow up with:
- Switch out the naive field parser with the DSL
- Switch out the naive query parser with discover utils. We can reuse the simplified model that metric extraction conversion code is using, since in this case we do not need to translate field expressions. So we just need the lexer + reconstruct operator precedence

File: src/sentry/sentry_metrics/querying/api.py (hunk 13 above)
Comment (iambriccardo) [reply src/sentry/sentry_metrics/querying/api.py:282]: I will change that
